cmd/mooshy/mooshy.go

```diff
@@ -174,6 +174,31 @@ func latestMoosh(token string) (string, error) {
 	return "", errors.New("not found")
 }
 
+func pollMagicNumber(conn net.Conn, port string, addr string, connected chan struct{}) {
+	var i int
+	var timeout time.Duration = 1 * time.Second
+	for i = 0; i < 5; i++ {
+		_, err := conn.Write([]byte(mooshy.MagicNumber + port))
+		if err != nil {
+			log.Fatalf("Failed to send magic number to %s: %s", addr, err)
+		} else if i == 0 {
+			log.Printf("Sent magic number to %s", addr)
+		} else {
+			log.Printf("Resent magic number (waiting %.0f sec)...", timeout.Seconds())
+		}
+
+		time.Sleep(timeout)
+		timeout *= 2
+
+		select {
+		case <-connected:
+			return
+		default:
+		}
+	}
+	log.Fatalf("Did not receive a connection after %d tries", i)
```

Comment:
Maybe return an error and let the caller decide if it should `exit` or a recovery is possible?

Reply:
Agreed. I changed the function to return an error instead.

Reply:
What's the best way to catch errors from a goroutine?

Reply:
Don't know if it's the best, but in general one of the ways would be:
```go
errch := make(chan error, 1)
go func() {
   errch <- fn()
}
<-errch // once this succeeds, the fn has returned
```

Here I guess you can still do the `log.Fatal`, but then inside a 
```go
go func() {
   if err := fn(); err != nil {
       log.Fatalf("Failed to...: %s", err)
   }
}
```

Keep in mind, that none of the `defer`ed statements execute after the `log.Fatal`.